### PR TITLE
st standalone model

### PR DESCRIPTION
**Pull request by AlexGrinch**
# What does this PR do ?

Adds support for ConformerEnc-TransformerDec speech translation (ST) model.

**Collection**: asr

# Changelog 
- Added baseline ST model
- Tested to work with both yttm and SentencePiece (SP) tokenizers, SP is considered default

# Usage
* You can potentially add a usage example below

```python
# Add a code snippet demonstrating how to use this 
```

# Before your PR is "Ready for review"
**Pre checks**:
- [ ] Make sure you read and followed [Contributor guidelines](https://github.com/NVIDIA/NeMo/blob/main/CONTRIBUTING.md)
- [ ] Did you write any new necessary tests?
- [ ] Did you add or update any necessary documentation?
- [ ] Does the PR affect components that are optional to install? (Ex: Numba, Pynini, Apex etc)
  - [ ] Reviewer: Does the PR have correct import guards for all optional libraries?
  
**PR Type**:
- [ ] New Feature
- [ ] Bugfix
- [ ] Documentation

If you haven't finished some of the above items you can still open "Draft" PR.

## Who can review?

Anyone in the community is free to review the PR once the checks have passed. 
[Contributor guidelines](https://github.com/NVIDIA/NeMo/blob/main/CONTRIBUTING.md) contains specific people who can review PRs to various areas.

# Additional Information
* Related to # (issue)

### CODE REVIEW

File: examples/asr/conf/transformer_dec/speech_translation_transf.yaml [hunk 1]
```diff
@@ -0,0 +1,235 @@
+# It contains the default values for training a Conformer-CTC ASR model, large size (~120M) with CTC loss and sub-word encoding.
```
Comment (titu1994) [inline examples/asr/conf/transformer_dec/speech_translation_transf.yaml:1]: Hmm wondering if it should be conf/speech_translation/fastconformer/speech_translation_transformer.yaml ?

File: examples/asr/conf/transformer_dec/speech_translation_transf.yaml [hunk 2]
```diff
@@ -0,0 +1,235 @@
...
+
+# Architecture and training config:
```
Comment (titu1994) [inline examples/asr/conf/transformer_dec/speech_translation_transf.yaml:3]: Update text here for ST

File: examples/asr/conf/transformer_dec/speech_translation_transf.yaml [hunk 3]
```diff
@@ -0,0 +1,235 @@
...
+  encoder:
+    _target_: nemo.collections.asr.modules.ConformerEncoder
+    feat_in: ${model.preprocessor.features}
+    feat_out: -1 # you may set it if you need different output size other than the default d_model
+    n_layers: 17
+    d_model: 512
+
+    # Sub-sampling params
+    subsampling: dw_striding # vggnet or striding, vggnet may give better results but needs more memory
+    subsampling_factor: 8 # must be power of 2
```
Comment (titu1994) [inline examples/asr/conf/transformer_dec/speech_translation_transf.yaml:100]: Is this fast conformer?

File: examples/asr/conf/transformer_dec/speech_translation_transf.yaml [hunk 4]
```diff
@@ -0,0 +1,235 @@
...
+    sched:
+      name: InverseSquareRootAnnealing
+      #d_model: ${model.encoder.d_model}
+      # scheduler config override
+      warmup_steps: 1000
+      warmup_ratio: null
+      min_lr: 1e-6
+
+# Initialize model encoder with pre-trained ASR FastConformer encoder for faster convergence and improved accuracy
+init_from_nemo_model:
```
Comment (titu1994) [inline examples/asr/conf/transformer_dec/speech_translation_transf.yaml:190]: This should go at the top before "model" section in config

File: examples/asr/speech_translation/speech_to_text_transf.py [hunk 5]
```diff
@@ -0,0 +1,75 @@
...
+import pytorch_lightning as pl
+from omegaconf import OmegaConf
+
+from nemo.collections.asr.models import EncDecTransfModelBPE
+from nemo.core.config import hydra_runner
+from nemo.utils import logging
+from nemo.utils.exp_manager import exp_manager
+
+
+@hydra_runner(config_path="../conf/transformer_dec/", config_name="speech_translation_transf_test")
```
Comment (titu1994) [inline examples/asr/speech_translation/speech_to_text_transf.py:57]: transf_test?

File: nemo/collections/asr/models/transformer_bpe_models.py [hunk 6]
```diff
@@ -0,0 +1,600 @@
...
+from nemo.collections.asr.parts.mixins import ASRBPEMixin
+from nemo.collections.common.losses import SmoothedCrossEntropyLoss
+from nemo.collections.common.metrics import GlobalAverageLossMetric
+from nemo.collections.common.parts import transformer_weights_init
+
+try:
+    from sacrebleu import corpus_bleu
+    from nemo.collections.nlp.modules.common import TokenClassifier
+    from nemo.collections.nlp.modules.common.lm_utils import get_transformer
+    from nemo.collections.nlp.modules.common.transformer import BeamSearchSequenceGenerator, TransformerEncoder
```
Comment (titu1994) [inline nemo/collections/asr/models/transformer_bpe_models.py:41]: These will soon be deprecated from NLP in coming releases. It might be better to either 
1) Move them to commons dir.
2) Copy them specifically into ASR domain.

Im more for (1) since TTS might need the modules as well eventually.

File: nemo/collections/asr/models/transformer_bpe_models.py [hunk 7]
```diff
@@ -0,0 +1,600 @@
...
+        if 'tokenizer' not in cfg:
+            raise ValueError("`cfg` must have `tokenizer` config to create a tokenizer !")
+
+        # Setup the tokenizer
+        self._setup_tokenizer(cfg.tokenizer)
+
+        super().__init__(cfg=cfg, trainer=trainer)
+
+        # Setup audio preprocessor
+        self.preprocessor = EncDecTransfModelBPE.from_config_dict(self._cfg.preprocessor)
```
Comment (titu1994) [inline nemo/collections/asr/models/transformer_bpe_models.py:83]: You can self.cfg instead of self._cfg

File: nemo/collections/asr/models/transformer_bpe_models.py [hunk 8]
```diff
@@ -0,0 +1,600 @@
...
+
+        # Setup audio encoder
+        self.encoder = EncDecTransfModelBPE.from_config_dict(self._cfg.encoder)
+
+        # Add projection layer if encoder and decoder differ in hidden size
+        if self._cfg.encoder['d_model'] != self._cfg.transf_decoder['hidden_size']:
+            self.adapter = torch.nn.Linear(self._cfg.encoder['d_model'], self._cfg.transf_decoder['hidden_size'])
+        else:
+            self.adapter = lambda x: x
+
```
Comment (titu1994) [inline nemo/collections/asr/models/transformer_bpe_models.py:93]: Dont use lambda, use nn.Identity()

File: nemo/collections/asr/models/transformer_bpe_models.py [hunk 9]
```diff
@@ -0,0 +1,600 @@
...
+                    beam_hypotheses = [self.tokenizer.ids_to_text(hyp) for hyp in beam_hypotheses]
+
+                    if return_hypotheses:
+                        # dump log probs per file
+                        for idx in range(logits.shape[0]):
+                            current_hypotheses[idx].y_sequence = logits[idx][: logits_len[idx]]
+
+                    hypotheses += beam_hypotheses
+
+                    del test_batch
```
Comment (titu1994) [inline nemo/collections/asr/models/transformer_bpe_models.py:254]: Maybe del other torch tensors that are not used until next batch as well

File: nemo/collections/asr/models/transformer_bpe_models.py [hunk 10]
```diff
@@ -0,0 +1,600 @@
...
+        """
+        has_input_signal = input_signal is not None and input_signal_length is not None
+        has_processed_signal = processed_signal is not None and processed_signal_length is not None
+        if (has_input_signal ^ has_processed_signal) == False:
+            raise ValueError(
+                f"{self} Arguments ``input_signal`` and ``input_signal_length`` are mutually exclusive "
+                " with ``processed_signal`` and ``processed_signal_len`` arguments."
+            )
+
+        if not has_processed_signal:
```
Comment (titu1994) [inline nemo/collections/asr/models/transformer_bpe_models.py:427]: Document which modules are being executed

File: nemo/collections/asr/models/transformer_bpe_models.py [hunk 11]
```diff
@@ -0,0 +1,600 @@
...
+
+        dec_mask = lens_to_mask(transcript_length, transcript.shape[1]).to(transcript.dtype)
+        dec_states = self.transf_decoder(
+            input_ids=transcript, decoder_mask=dec_mask, encoder_embeddings=enc_states, encoder_mask=enc_mask
+        )
+        transf_log_probs = self.log_softmax(hidden_states=dec_states)
+
+        return transf_log_probs, encoded_len, enc_states, enc_mask
+
+    def compute_audio_loss(self, batch):
```
Comment (titu1994) [inline nemo/collections/asr/models/transformer_bpe_models.py:451]: Isnt audio loss originally ctc ?

File: nemo/collections/asr/models/transformer_bpe_models.py [hunk 12]
```diff
@@ -0,0 +1,600 @@
...
+    batch_size = lens.shape[0]
+    mask = torch.arange(max_length).repeat(batch_size, 1).to(lens.device) < lens[:, None]
+    return mask
+
+
+class EncDecTransfModelBPE(ASRModel, ExportableEncDecModel, ASRBPEMixin):
+    """Base class for encoder decoder CTC-based models."""
+
+    def __init__(self, cfg: DictConfig, trainer: Trainer = None):
+
```
Comment (titu1994) [inline nemo/collections/asr/models/transformer_bpe_models.py:73]: Check if NLP_AVAILABLE or not, raise error otherwise with appropriate warning

File: nemo/collections/asr/parts/mixins/mixins.py [hunk 13]
```diff
@@ -83,7 +83,7 @@ def _setup_monolingual_tokenizer(self, tokenizer_cfg: DictConfig):
                 with open_dict(self.cfg.tokenizer):
                     self.cfg.tokenizer.hf_kwargs = tokenizer_cfg.get('hf_kwargs')
 
-        if self.tokenizer_type not in ['bpe', 'wpe']:
+        if self.tokenizer_type not in ['bpe', 'wpe', 'yttm']:
```
Comment (titu1994) [inline nemo/collections/asr/parts/mixins/mixins.py:86]: Dont support yttm its mostly deprecated. https://github.com/NVIDIA/NeMo/issues/6854

Comment (titu1994) [reply nemo/collections/asr/parts/mixins/mixins.py:86]: Sentencepiece with spe mode does just the same thing as yttm

**Review by titu1994 — CHANGES_REQUESTED** [review]
Overall it looks great, minor comments mostly. One thing is do not use yttm, its support is flaky.

File: nemo/collections/asr/models/transformer_bpe_models.py (hunk 6 above)
Comment (AlexGrinch) [reply nemo/collections/asr/models/transformer_bpe_models.py]: I plan to move every TTS tokenisation dependency and NLP-related tokenizer utils to common.tokenizers. However, it should be separate PR with discussion as it might affect some of the already released models.

File: nemo/collections/asr/models/transformer_bpe_models.py [hunk 14]
```diff
@@ -0,0 +1,601 @@
...
+        if hasattr(self.cfg, 'spec_augment') and self.cfg.spec_augment is not None:
+            self.spec_augmentation = EncDecTransfModelBPE.from_config_dict(self.cfg.spec_augment)
+        else:
+            self.spec_augmentation = None
+
+        self.val_loss = GlobalAverageLossMetric(dist_sync_on_step=False, take_avg_loss=True)
+
+    @torch.no_grad()
+    def transcribe(
```
Comment (titu1994) [inline nemo/collections/asr/models/transformer_bpe_models.py:174]: Should probably be called translate ? And you can add a forwarding from transcribe to the actual translate() function.

**Review by titu1994 — APPROVED** [review]
Looks great ! Nice addition to NeMo ~

File: examples/asr/conf/speech_translation/fast-conformer_transformer.yaml
Comment (titu1994) [inline examples/asr/conf/speech_translation/fast-conformer_transformer.yaml:1]: speech_translation/fast_conformer/fastconformer_transformer.yaml

File: nemo/collections/asr/models/transformer_bpe_models.py (hunk 12 above)
Comment (titu1994) [reply nemo/collections/asr/models/transformer_bpe_models.py:74]: ++

**Review by titu1994 — COMMENTED** [review]
Overall it looks great.
Two things - need to change file name and add check for NLP available or not and crash at init if sacrebleu is not available.

**Review by titu1994 — APPROVED** [review]

**Comment by titu1994** [comment]
Also needs documentation + tutorial @AlexGrinch
